### PR TITLE
Adding GA to workshop

### PR DESCRIPTION
**Pull request by Jimmy-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
see #1391 
##### What are the significant changes in functionality due to this change request?
Adding Genetic Algorithms to worshop

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: ravenframework/Optimizers/GeneticAlgorithm.py [hunk 1]
```diff
@@ -611,7 +613,7 @@ def flush(self):
 
   # END queuing Runs
   # * * * * * * * * * * * * * * * *
-  def _resolveNewGeneration(self, traj, rlz, objectiveVal, fitness, info):
+  def _resolveNewGeneration(self, traj, rlz, objectiveVal, fitness, g, info):
```
Comment (wangcj05) [inline ravenframework/Optimizers/GeneticAlgorithm.py:616]: add docstring for "g"

File: ravenframework/Optimizers/RavenSampled.py [hunk 2]
```diff
@@ -642,7 +642,7 @@ def _updateSolutionExport(self, traj, rlz, acceptable, rejectReason):
     toExport.update(self.denormalizeData(dict((var, rlz[var]) for var in self.toBeSampled)))
     # constants and functions
     toExport.update(self.constants)
-    toExport.update(dict((var, rlz[var]) for var in self.dependentSample))
+    # toExport.update(dict((var, rlz[var]) for var in self.dependentSample))
```
Comment (wangcj05) [inline ravenframework/Optimizers/RavenSampled.py:645]: Let's put a if check here:
```Python
toExport.update(dict((var, rlz[var]) for var in self.dependentSample if var in rlz))
```

File: tests/framework/AnalyticModels/optimizing/projectileConstraints.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# @ author: Mohammad Abdo (@Jimmy-INL)
+
+def implicitConstraint(Input):
+  """
+    Evaluates the constraint function @ a given point ($\vec(x)$)
+    @ In, self, object, RAVEN container
```
Comment (wangcj05) [inline tests/framework/AnalyticModels/optimizing/projectileConstraints.py:19]: update "self" to "Input"

File: tests/framework/AnalyticModels/optimizing/projectileConstraints.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            3) f(x,y) <= b then g = b - f
+            4) f(x,y)  = c then g = 0.001 - (f(x,y) - c)
+  """
+  g = 250 - Input.ymax
+  return g
+
+def constrain(self,Input):
+  """
+    Constrain calls the constraint function.
+    @ In, self, object, RAVEN container
```
Comment (wangcj05) [inline tests/framework/AnalyticModels/optimizing/projectileConstraints.py:36]: update "self" to "Input"

File: tests/framework/AnalyticModels/optimizing/projectileConstraints.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    @ Out, explicitConstrain, float, positive if the constraint is satisfied
+           and negative if violated.
+  """
+  implicitConstraint = implicitConstraint(Input)
```
Comment (wangcj05) [inline tests/framework/AnalyticModels/optimizing/projectileConstraints.py:40]: why both implicitConstraint and "Constraint" in the same file, I guess you will get some problem with it.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
I have a couple of comments. In addition,
1. Please open an issue for this PR
2. Please open another issue regarding the handling of solution export variables 
3. Please open an issue regarding dependent variables

**Comment by Jimmy-INL** [comment]
Screen Shot 2022-07-25 at 9 54 53 PM

**Comment by Jimmy-INL** [comment]
Screen Shot 2022-07-25 at 10 12 33 PM

**Comment by Jimmy-INL** [comment]
Screen Shot 2022-07-25 at 10 14 40 PM

**Comment by Jimmy-INL** [comment]
Screen Shot 2022-07-25 at 10 18 32 PM

**Comment by Jimmy-INL** [comment]
these are all the timed out tests

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
PR looks good, tests passed on local machine. Regression test system is not available this week, PR will be retested once the test system is back. This PR is required for RAVEN workshop. As CCB, I approve this PR.

**Pull request merged**